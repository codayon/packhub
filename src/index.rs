use std::io::Write;

use chrono::Utc;
use libflate::gzip::{Encoder, HeaderBuilder, EncodeOptions};
use askama::Template;

use crate::{detect::Package, deb::DebAnalyzer};

pub struct AptIndices<'a> {
    data: &'a [u8],
    package: &'a Package,
}

#[derive(Template)]
#[template(path = "Release")] 
struct ReleaseIndex {
    date: String,
    files: Vec<Files>,
}

#[derive(Template)]
#[template(path = "Packages")] 
struct PackageIndex<'a> {
    control: &'a str,
    sum256: String,
    size: usize,
    filename: String,
}

struct Files {
    sum256: String,
    size: usize,
    path: String, 
}

impl<'a> AptIndices<'a> {
    pub fn new(package: &'a Package, data: &'a [u8]) -> Self {
        AptIndices { data, package }
    }

    pub fn get_package_index(&self) -> String {
        let deb = DebAnalyzer::new(self.data);

        let control = deb.get_control_data().trim_end();
        let filename= format!("pool/stable/{}/{}", self.package.version(), self.package.file_name());
        let size = self.data.len();
        let sum256 = sha256::digest(self.data);

        let index = PackageIndex { control, sum256, size, filename };
        index.render().unwrap()
    }

    pub fn get_release_index(&self) -> String {
        let date = Utc::now().to_rfc2822();

        let packages = self.get_package_index();
        let packages = packages.as_bytes();

        let packages_gz = gzip_compression(packages);

        let files = vec![
            Files { sum256: sha256::digest(packages), size: packages.len(), path: "main/binary-amd64/Packages".to_string() },
            Files { sum256: sha256::digest(&packages_gz), size: packages_gz.len(), path: "main/binary-amd64/Packages.gz".to_string() }
        ];

        let index = ReleaseIndex { date, files };

        index.render().unwrap()
    }
}

pub fn gzip_compression(data: &[u8]) -> Vec<u8> {
    let header = HeaderBuilder::new().modification_time(0).finish();
    let options = EncodeOptions::new().header(header);
    let mut encoder = Encoder::with_options(Vec::new(), options).unwrap();
    encoder.write_all(data).unwrap();

    let gzip = encoder.finish();
    let gzip = gzip.into_result().unwrap();

    gzip
}

#[cfg(test)]
mod tests {
    use std::fs;

    use insta::assert_snapshot;

    use super::*;

    #[test]
    fn test_apt_indices() {
        let package = Package::detect_package("OpenBangla-Keyboard_2.0.0-ubuntu20.04.deb", "2.0.0".to_owned(), "https://github.com/OpenBangla/OpenBangla-Keyboard/releases/download/2.0.0/OpenBangla-Keyboard_2.0.0-ubuntu20.04.deb".to_owned()).unwrap();
        let data = fs::read("data/OpenBangla-Keyboard_2.0.0-ubuntu20.04.deb").unwrap();

        let indices = AptIndices::new(&package, &data);

        // Packages
        let packages = indices.get_package_index();
        assert_snapshot!(packages);
        
        // Release
        let release = indices.get_release_index();
        insta::with_settings!({filters => vec![
            // Date is a changing value, so replace it with a hardcoded value.
            (r"Date: .+", "Date: [DATE]"),
        ]}, {
            assert_snapshot!(release);
        });
    }
}
