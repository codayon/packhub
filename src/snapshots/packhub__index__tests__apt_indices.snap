---
source: src/index.rs
expression: packages
---
Architecture: amd64
Depends: ibus (>= 1.5.1), libc6 (>= 2.29), libgcc-s1 (>= 4.2), libglib2.0-0 (>= 2.12.0), libibus-1.0-5 (>= 1.5.1), libqt5core5a (>= 5.12.2), libqt5gui5 (>= 5.0.2) | libqt5gui5-gles (>= 5.0.2), libqt5network5 (>= 5.0.2), libqt5widgets5 (>= 5.0.2), libstdc++6 (>= 5.2), libzstd1 (>= 1.3.2)
Description: OpenSource Bengali input method
 OpenBangla Keyboard is an OpenSource, Unicode compliant Bengali Input Method for GNU/Linux systems. It&apos;s a full fledged Bengali input method with typing automation tools, includes many famous typing methods such as Avro Phonetic, Probhat, Munir Optima, National (Jatiya) etc.
 .
 Most of the features of Avro Keyboard are present in OpenBangla Keyboard. So Avro Keyboard users will feel at home with OpenBangla Keyboard in Linux.
 .
Homepage: https://openbangla.github.io/
Maintainer: OpenBangla Team <openbanglateam@gmail.com>
Package: openbangla-keyboard
Priority: optional
Section: utils
Version: 2.0.0
Installed-Size: 12263
SHA256: d368f56f14d07d0f32482d9e56e68750ed1b4f68157fc53f6574c6fde6c4457e
Size: 5332906
Filename: pool/stable/2.0.0/OpenBangla-Keyboard_2.0.0-ubuntu20.04.deb
